pdns/pdns_recursor.cc

```diff
@@ -2908,6 +2910,10 @@ static void houseKeeping(void *)
       if(!((cleanCounter++)%40)) {  // this is a full scan!
 	time_t limit=now.tv_sec-300;
         SyncRes::pruneNSSpeeds(limit);
+        limit = now.tv_sec - SyncRes::s_serverdownthrottletime * 10;
+        SyncRes::pruneFailedServers(limit);
+        limit = now.tv_sec - 2*3600;
+        SyncRes::pruneEDNSStatuses(limit);
```

Comment:
I'm wondering if we can have an issue due to the fact that `SyncRes::asyncresolveWrapper()` keeps a pointer to an entry of the map during the call to `asyncresolve()`, which will yield the control back to the mthread kernel. I'm pretty sure we have a race condition if the map is cleaned up at that time, and the pointer is no longer valid when `asyncresolve()` returns.
We should likely do a lookup again when we return instead.

Reply:
Indeed, the same thought occurred to me this morning. Will redraft.